File: 00_Arachnida/Spider/srcs/Spider.cpp
#include <iostream>
#include "Spider.hpp"

const string	Spider::get_url() const {
	return this->url;
}

Spider::Spider(int ac, char **av) : path("./data/"), depth(0), done({}) {
	string	opt;
	string	arg;

	for (int i = 1; i < ac; i++) {
		arg = av[i];

		if (arg[0] == '-') {
			if (arg.length() != 2)
				throw invalid_argument("Bad option call");
			if (arg[1] == 'r') {
				if (opt.find("r") != string::npos)
					throw invalid_argument("Duplication option r");
				opt += "r";
				if (opt.find("l") == string::npos)
					depth = 5;
			} else if (arg[1] == 'l') {
				if (opt.find("l") != string::npos)
					throw invalid_argument("Duplication option l");
				opt += "l";
				depth = stoi(av[++i]);
			} else if (arg[1] == 'p') {
				if (opt.find("p") != string::npos)
					throw invalid_argument("Duplication option p");
				opt += "p";
				path = av[++i];
			} else {
				throw invalid_argument("Wrong options");
			}
		} else {
			if (i != ac - 1)
				throw invalid_argument("URL should be last");
			url = av[i];
		}
	}
	this->check(opt);
}

Spider::~Spider() {}

void	Spider::check(string opt) {

	// Check if options validity
	if (opt.find("l") != string::npos && opt.find("r") == string::npos)
		throw invalid_argument("Depth option without recursive");
	if (opt.find("l") != string::npos && (depth < 0 || depth > 1000))
		throw invalid_argument("Invalid depth (must be between 0 and 1000)");
	// Check for folder path validity
	int	check = mkdir(this->path.c_str(), 0777);
	if (this->path.back() != '/')
		this->path += "/";
	if (check < 0 && errno != EEXIST)
		throw invalid_argument("Unable to create directory \"" + this->path + "\"");
	// Check if URL is valid
	if (url.empty())
    	throw invalid_argument("No URL");
	// CURL *curl = curl_easy_init();
	// if(curl) {
	// 	curl_easy_setopt(curl, CURLOPT_URL, this->url.c_str());
	
	// 	CURLcode res = curl_easy_perform(curl);
	// 	if (res != CURLE_OK) {
	// 		curl_easy_cleanup(curl);
	// 		throw std::invalid_argument("URL invalid or not responding");
	// 	}
	// 	curl_easy_cleanup(curl);
	// }
}


void	Spider::scrap(const string url) {
	vector<string>	links;
	vector<string>	imgs;
	string			content;

	if (find(this->done.begin(), this->done.end(), url) != this->done.end())
		return ;
	// Get the page
	CURL *curl = curl_easy_init();
	if(!curl)
		throw exception();
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_to_string);
	curl_easy_setopt(curl, CURLOPT_WRITEDATA, &content); 
    CURLcode res = curl_easy_perform(curl);
	
	if (res != CURLE_OK)
		throw runtime_error(url + " is invalid or not responding");

	curl_easy_cleanup(curl);

	// Parse to get each image url
	imgs = parse_get_all(content, "//img[@src]", "src");

	// Download every images
	for (int i = 0; i < (int)imgs.size(); i++) {
		this->download_img(imgs[i]);
	}

	// Push url to done list
	this->done.push_back(url);

	if (this->depth > 0) {
		// Parse to get each link in the page
		links = parse_get_all(content, "//a[@href]", "href");
		this->depth--;
		// Scrap every links
		for (int i = 0; i < (int)links.size(); i++) {
			this->scrap(links[i]);
		}
	}
}

void		Spider::download_img(const string url) {
	string	filename;

	filename = get_img_name(url);
	if (filename.empty())
		return ;
	
	filename = this->path + filename;
    CURL* curl = curl_easy_init();
    if (curl) {
        FILE* fp = fopen(filename.c_str(), "wb");
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
        CURLcode res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        fclose(fp);

        if (res != CURLE_OK)
            throw runtime_error(url + "Could not download: " + url);
	}
}

File: 00_Arachnida/Spider/srcs/spider.py
import utils

import os
import requests
import shutil
import urllib
from urllib.parse import urlparse
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.chrome.options import Options

class Spider:
	imgs = 0

	def __init__(self, args):
		self.url = ""
		self.path = "./data/"
		self.depth = 0
		self.done = []

		opt = ""
		i = 0
		while i < len(args):
			if (args[i][0] == '-'):
				if len(args[i]) != 2:
					raise ValueError("Bad option call")
				
				match args[i][1]:
					case 'r':
						if 'r' in opt:
							raise ValueError("Duplication option r")
						opt += "r"
						if 'l' not in opt:
							self.depth = 5
					case 'l':
						if 'l' in opt:
							raise ValueError("Duplication option l")
						opt += "l"
						i += 1
						self.depth = int(args[i])
					case 'p':
						if 'p' in opt:
							raise ValueError("Duplication option p")
						opt += "p"
						i += 1
						self.path = args[i]
					case _:
						raise ValueError("Wrong option")
			else:
				if i < len(args) - 1:
					raise ValueError("Bad usage")
				self.url = args[i]
			i += 1

		self.check(opt)
		''' create session '''
		options = Options()
		options.add_argument("--headless")
		self.driver = webdriver.Chrome(options=options)


	def check(self, opt):
		''' Check for option validity '''
		if 'l' in opt and 'r' not in opt:
			raise ValueError("Depth option without recursive")
		if 'l' in opt and (self.depth < 0 or self.depth > 10):
			raise ValueError("Invalid depth (must be between 0 and 10)")
		
		''' Check for folder path '''
		if os.path.isdir(self.path):
			shutil.rmtree(self.path)
		os.makedirs(self.path)
		if self.path[-1] != '/':
			self.path += "/"

		''' Check for valid url '''
		parsed = urlparse(self.url)
		if not parsed.scheme or not parsed.netloc:
			raise ValueError("Invalid url")

	def scrap(self, url, depth=0):
		elements = []
		links = []

		if url in self.done:
			return

		print(f"Downloading images from: {url} (depth: {self.depth - depth}) ...\n")

		''' Get the page content '''
		self.driver.get(url)
		WebDriverWait(self.driver, 5).until(
            lambda d: d.execute_script("return document.readyState") == "complete"
        )
		if not self.driver.current_url:
			return

		''' Parse and store each image url '''
		elements = self.driver.find_elements(By.TAG_NAME, "img")
		links = [e.get_attribute("src") for e in elements]

		''' Download each images '''
		headers = {
            "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115 Safari/537.36",
            "Referer": url,
        }
		for x in links:
			filename = utils.get_img_name(x)
			if not filename or x == None:
				continue
			filepath = self.path + filename
			img_url = urllib.parse.urljoin(url, x)
			with open(filepath, 'wb') as file:
				file.write(requests.get(img_url, headers=headers).content)
				Spider.imgs += 1

		self.done.append(url)
		elements.clear()
		links.clear()

		''' Scrap all the links in the page '''
		if depth > 0:
			elements = self.driver.find_elements(By.TAG_NAME, "a")
			links = [e.get_attribute("href") for e in elements]
			for x in links:
				if x != None and not x.startswith("#") and not x.startswith("mailto:") \
					and not x.startswith("javascript"):
					link_url = urllib.parse.urljoin(self.url, x)
					self.scrap(link_url, depth - 1)

	def __del__(self):
		driver = getattr(self, "driver", None)
		if driver:
			driver.quit()

	def	print(self):
		print("--- SPIDER ---\n\n")
		print(f"* url: {self.url}\n\n")
		print(f"* path: {self.path}\n\n")
		print(f"* depth: {self.depth}\n\n")
		print("* done:\n")
		for x in self.done:
			print(f"    - {x}\n")